### DISCUSSION TITLE
Using motoros2 with rviz

### DISCUSSION THREAD

**Discussion by yongpaulsit** [Yaskawa-Global/motoros2#251] [Q&A] [answered]
Hi, 
I'm using a YRC1000micro and Motomini, and ROS2 Humble on Ubuntu 22.04. 
Following the readme, I was able to setup MotoROS2 on the YRC1000micro and wrote a client to use the Actions and Services to perform simple point to point movement. 

Now I'm trying to visualize the Motomini in rviz. 
I was able to simulate a Motomini using the [motoman](https://github.com/gavanderhoorn/motoman/tree/ros2) package from @gavanderhoorn and following the ROS2 tutorial on [Using URDF with robot_state_publisher](https://docs.ros.org/en/humble/Tutorials/Intermediate/URDF/Using-URDF-with-Robot-State-Publisher.html). 

I am aware that:
- as of ROS2, the joint names from the motoman_motomini_support package need to be renamed e.g.) 'joint_1_s' to 'joint_1'.
- the motoros2 driver installed on the robot controller (YRC1000micro) is publishing to /tf /joint_states and /robot_status so robot_state_publisher should not be needed

Using rviz, I am able to use /tf to visualize the Motomini, but I am unable to get it working together with the URDF.
I noticed that when using robot_state_publisher and joint_state_publisher to simulate the Motomini, the data published to /tf is different.
Is there a way to solve this?

**Comment by gavanderhoorn** [comment]
> * as of ROS2, the joint names from the motoman_motomini_support package need to be renamed e.g.) 'joint_1_s' to 'joint_1'.

it's not actually a requirement, but it would be good to migrate to the non-OEM-specific names, yes. But whether you can will also depend on what the URDF uses/specifies.

MotoROS2 allows you to configure the names of all joints, so it doesn't force you to use `joint_1`. See [Can names of joints be changed?](https://github.com/Yaskawa-Global/motoros2/blob/0cf1110c520dbf7dfe421cf852954ad562d847c6/doc/faq.md#can-names-of-joints-be-changed) in the FAQ.

> * the motoros2 driver installed on the robot controller (YRC1000micro) is publishing to `/tf`, `/joint_states` and `/robot_status` so `robot_state_publisher` should not be needed

to clarify: `/robot_status` carries [this message](https://github.com/ros-industrial/industrial_core/blob/ea3d805e1271550be312409d6474177e67796e15/industrial_msgs/msg/RobotStatus.msg). It encodes information about the state of the Yaskawa controller (fi: are the servos turned on, is the robot currently moving).

The `robot_state_publisher` node does not consume nor produce these messages. It is used to perform FK (forward kinematics) by combining information in the URDF with the `JointState` messages (which it takes from the `/joint_states` topic by default).

To summarise: I understand your confusion, but you still definitely need the `robot_state_publisher`. The `robot_status` topic is unrelated to that.

> Using rviz, I am able to use /tf to visualize the Motomini, but I am unable to get it working together with the URDF.

The TF frames broadcast by MotoROS2 will not be sufficient for this purpose. You do absolutely (still) need the `robot_state_publisher`.

I expect your robot will appear as soon as you start a `robot_state_publisher` and provide it with the MotoMini's URDF.

You could take a look at [this launch file](https://github.com/gavanderhoorn/motoman/blob/302fc2309f17de3626df536f53ef44ee17238c7f/motoman_gp25_support/launch/view_gp25.launch). You should essentially only need to remove / comment [this line](https://github.com/gavanderhoorn/motoman/blob/302fc2309f17de3626df536f53ef44ee17238c7f/motoman_gp25_support/launch/view_gp25.launch#L8) for the `robot_state_publisher` in that launch file to take the `JointState` messages coming from MotoROS2 instead of the 'virtual' ones published by the `joint_state_publisher_gui`.

> I noticed that when using robot_state_publisher and joint_state_publisher [..], the data published to /tf is different. Is there a way to solve this?

As I wrote, MotoROS2 publishes different frames, which are not intended for visualisation purposes. The fact that tree and the one you noticed when using just the RSP and the JSP are different is not what caused the issue you described.

And pedantic, but:

> I noticed that when using robot_state_publisher and joint_state_publisher to simulate the Motomini, the data published to /tf is different. Is there a way to solve this?

With a `robot_state_publisher` and a `joint_state_publisher` you're not so much *simulating* something, but exercising the FK functionality I described earlier. It allows you to visualise the robot model using `JointState` messages published by the `joint_state_publisher` (likely the `joint_state_publisher_gui` node).

**Reply by yongpaulsit** [reply, marked as answer]
Thanks for answering. I was having trouble understanding why rviz was only using the `/tf` from `motoros2` and not from ~`joint_state_publisher`~ edit: `robot_state_publisher`, after reading your response I looked into the `motoros2_config.yaml` file and found that I can disable `motoros2` publishing to `/tf`. I set `publish_tf` to `false` and changed the joint names to follow the `joint_1_s` format (not sure if this was necessary). 

my `motoros2_config.yaml`
```yaml
publish_tf: false
joint_names: [joint_1_s, joint_2_l, joint_3_u, joint_4_r, joint_5_b, joint_6_t]
```

After updating the config file on YRC1000micro, rviz now shows the Motomini in real time.
![screenshot](https://github.com/Yaskawa-Global/motoros2/assets/92104761/188a7787-0e3a-4d7a-9aa9-668004c07ad9)

Connecting it to my client, I can send joint trajectories.
![untitled](https://github.com/Yaskawa-Global/motoros2/assets/92104761/0624f8f1-0cad-4481-ab97-320309641d5a)

rqt_graph for reference:
![rosgraph20](https://github.com/Yaskawa-Global/motoros2/assets/92104761/71bb89ad-8cac-46d6-9f2d-0cf407e56edc)

**Reply by gavanderhoorn** [reply]
Glad you got things to work.

> I was having trouble understanding why rviz was only using the `/tf` from `motoros2` and not from `joint_state_publisher`, [..]

Just to clarify: the JSP does not broadcast any TF. That's solely the job of the `robot_state_publisher`.

And RViz can't really separate different TF sources. It will consume *all* TF frames broadcast by all broadcasters (by default).

> [..] after reading your response I looked into the `motoros2_config.yaml` file and found that I can disable `motoros2` publishing to `/tf`.

yes, that is possible.

It's not required for what you want(ed) to do, but it's supported.

> I set `publish_tf` to `false` and changed the joint names to follow the `joint_1_s` format (not sure if this was necessary).

this would only be necessary if you're using a URDF with those joint names.

> Connecting it to my client, I can send joint trajectories. 

that's very nice indeed.

your client looks interesting.

Would you happen to have that shared anywhere?

**Reply by yongpaulsit** [reply]
> Just to clarify: the JSP does not broadcast any TF. That's solely the job of the robot_state_publisher.

I just realized that I kept saying `joint_state_publisher` when I meant `robot_state_publisher`, thanks for clarifying.

> Would you happen to have that shared anywhere?

I'm currently using [python tkinter](https://docs.python.org/3/library/tkinter.html) to (in a way,) map the motoros2 API to a GUI. The current interface is something I made to test the API and reinforce my learning of ROS2 and programming, and definitely not in a state that I have confidence in sharing. I'm still learning to use the tkinter library, and my goal it to create an interface similar to that of MotoCom32 or MotoComES before I share it on github.

![client](https://github.com/Yaskawa-Global/motoros2/assets/92104761/0773d786-1ac6-4f16-af0d-9eabcc5381f4)
![motocom32](https://github.com/Yaskawa-Global/motoros2/assets/92104761/90f4c63a-8af6-4cc3-b398-fa30d597b4c2)

**Comment by i435f34** [comment]
Hello @yongpaulsit, 
I'm pretty much trying to do exactly what you're doing but I'm having a bit of trouble since I don't have much experience with ROS. Do you have by any chance some code to share to be able to control a Motomini and visualize it in Rviz ? 
I would also be glad to have a look at your tkinter GUI if you feel comfortable sharing it.
 
Thanks a lot
